### PR TITLE
enable spotless formatter

### PR DESCRIPTION
**Pull request by J-N-K**
- apply "mvn spotless:apply"
- enable spotless checks by default
- exclude itest-inputs for feed

Supersedes #6874 

Signed-off-by: Jan N. Klug <jan.n.klug@rub.de>

### CODE REVIEW

**Comment by cpmeister** [comment]
It is very likely that this would cause conflicts with work that has not been merged yet.  Is there any way that this could be run automatically as part of git merging?

**Comment by cpmeister** [comment]
Actually, the addon migration to 3.0 would be the perfect time to do this.

**Comment by J-N-K** [comment]
I don‘t think so. The conflicts will be always present. And doing it incrementally is not a good idea either, if we don‘t enforce the code style, the risk of introducing wrong formatting again is near 100%.

**Comment by mhilbush** [comment]
There's something I don't understand. I use Eclipse for development. And, when I touch an xml file, I almost always use the Eclipse formatter. But, since there are bindings in here that I developed and/or work on, including one that was very recently merged, I can assume that a)  Eclipse formats differently than spotless, or b) my Eclipse config is borked up somehow such that it formats differently than Eclipse.

Fortunately, I have only one PR that's in the queue that's affected by this. If I had more, I guess I'd be a little more annoyed.

**Comment by J-N-K** [comment]
Are you actively triggering the XML formatter? In contrast to java files, XML files are not automatically formatted on save. And TTBOMK this cannot be configured.

**Comment by mhilbush** [comment]
Yes, I do the Shift-Alt-f, or whatever it is, to trigger it.

File: bundles/org.openhab.binding.insteon/src/main/resources/msg_definitions.xml [hunk 1]
```diff
@@ -1,410 +1,399 @@
...
-	<!--
-	//    Please keep messages ordered by increasing command number!
-	//
-	// The header is not an official Insteon concept. The boundary
-	// between header and message is the point where enough information
-	// is available to determine the full length of the incoming message.
-	// Sometimes that's just two bytes, sometimes 6 or 9.
-	//
-	->
+	<!-- // Please keep messages ordered by increasing command number! // // The header is not an official Insteon concept.
```
Comment (robnielsen) [inline bundles/org.openhab.binding.insteon/src/main/resources/msg_definitions.xml:2]: This didn't format correctly

File: bundles/org.openhab.binding.insteon/src/main/resources/device_types.xml [hunk 2]
```diff
@@ -1,580 +1,554 @@
...
-        <feature name="manualchangebuttonc">ManualChangeButton3</feature>
-        <feature name="manualchangebuttond">ManualChangeButton4</feature>
-        <feature name="manualchangebuttone">ManualChangeButton5</feature>
-        <feature name="manualchangebuttonf">ManualChangeButton6</feature>
-        <feature name="manualchangebuttong">ManualChangeButton7</feature>
-        <feature name="manualchangebuttonh">ManualChangeButton8</feature>
-     </feature_group>
-     <feature name="lastheardfrom">GenericLastTime</feature>
- </device>
+	<!-- device types # # PLEASE KEEP PRODUCT KEYS IN INCREASING ORDER: # # - first the devices with insteon assigned product
```
Comment (robnielsen) [inline bundles/org.openhab.binding.insteon/src/main/resources/device_types.xml:2]: This didn't format correctly

File: bundles/org.openhab.binding.insteon/src/main/resources/msg_definitions.xml (hunk 1 above)
Comment (J-N-K) [reply bundles/org.openhab.binding.insteon/src/main/resources/msg_definitions.xml:2]: Maybe we should exclude comments in XML from formatting.

Comment (wborn) [reply bundles/org.openhab.binding.insteon/src/main/resources/msg_definitions.xml:2]: I hope that will also stop it from messing up license headers on XML files. We saw that happen when applying it in openhab-core.

**Comment by wborn** [comment]
I think it doesn't really matter when it's applied. It helps to reduce diffs in reviews due to autosave actions if someone touches code that wasn't properly formatted. If it causes conflicts on big PRs, it will probably help if you also apply spotless on your branch

**Comment by theater** [comment]
Can we have this also updated on the development environment setup
documentation?
Are there XML formatter settings for eclipse which are compatible with this
that can be used by each binding developer locally instead of calling maven?

**Comment by J-N-K** [comment]
We wowuld need to change the XML-formatter in the sat-repo. This can't be fixed before OH3 because 0.10.0-SNAPSHOT of the SAT repo already contains OH3-only changes.

**Comment by wborn** [comment]
You can also add a file with the right the XML formatting options to this repo, see the [documentation](https://github.com/diffplug/spotless/blob/master/plugin-maven/README.md#applying-eclipse-wtp-to-css--html--etc).

**Comment by J-N-K** [comment]
@wborn, that may solve the XML problem, but there are other code-style changes (e.g. https://github.com/openhab/static-code-analysis/commit/1d154f8b149eca630146502253b36e369be8837f) and I don't think we should add all those changes to the addons-repo.

**Comment by wborn** [comment]
Another option could be that we release a new SAT version and only change the [dependency for spotless](https://github.com/openhab/openhab-addons/blob/41ee4ef9ce30e28b89487dd8f43855a9a3ec6c84/pom.xml#L601) to use the newer version.

File: bundles/org.openhab.binding.insteon/src/main/resources/msg_definitions.xml (hunk 1 above)
Comment (wborn) [reply bundles/org.openhab.binding.insteon/src/main/resources/msg_definitions.xml:2]: I found that https://github.com/openhab/static-code-analysis/pull/375 fixes the license formatting issues in SAT 0.10.0-SNAPSHOT though the plugin configuration was missing so I've fixed that with https://github.com/openhab/openhab-core/pull/1442.

We probably want to apply something similar for all XML files, so you could test that by adding

`formatCommentJoinLines=false`

if that doesn't help, all XML comment formatting can be disabled with:

`formatCommentText=false`

**Comment by J-N-K** [comment]
Wow. I'm impressed. I'm using a windows machine.

- `mvn spotless:apply` in the base directory formats and changes all line-endings in formatted files from `CRLF` to `LF`
- `mvn spotless:apply -rf :org.openhab.binding.airquality` in the base directory formats and changes line-endings in formatted files from `CRLF` to `LF`
- `mvn spotless:apply -pl bundles\org.openhab.binding.airquality` in the base directory formats and and correctly sets all line-endings to `CRLF`
- `mvn spotless:apply` in the project directory formats and correctly sets all line-endings to `CRLF`

I wouldn't care because this is fixed by git during commit, but the problem is that `mvn clean install` fails because of wrong line-endings...

**Comment by J-N-K** [comment]
It seems upgrading formatters to the latest versions and setting `core.autocrlf=true` (what already was there before) and `core.eol=crlf` in the git-config finally fixed it. According to the git documentation `core.eol` is ignored if `core.autocrlf=true` is set. But spotless seems to think differently...

I prepared a PR for SAT, the no-comment-join seems to have fixed the formatting issues in XMLs. We need a SAT release then and I can update the version here.

File: bundles/org.openhab.binding.insteon/src/main/resources/msg_definitions.xml (hunk 1 above)
Comment (robnielsen) [reply bundles/org.openhab.binding.insteon/src/main/resources/msg_definitions.xml]: This appears to be fixed in the latest commit.

File: bundles/org.openhab.binding.insteon/src/main/resources/device_types.xml (hunk 2 above)
Comment (robnielsen) [reply bundles/org.openhab.binding.insteon/src/main/resources/device_types.xml]: This appears to be fixed in the latest commit.

**Comment by wborn** [comment]
With those changes it works fine for me on Linux too. :+1: 

It does reformat the following file to Linux line endings:

`bundles/org.openhab.binding.amazonechocontrol/src/main/java/org/openhab/binding/amazonechocontrol/internal/WebSocketConnection.java`

It has some non-standard characters in it:

https://github.com/openhab/openhab-addons/blob/bfcebf06580fcfb92754b47c611019f0f0cfcb9d/bundles/org.openhab.binding.amazonechocontrol/src/main/java/org/openhab/binding/amazonechocontrol/internal/WebSocketConnection.java#L558

It works well on openhab-core too.

We should try to convince @kaikreuzer to release SAT 0.10.0 after also merging my 2 open PRs of course. :wink:

**Comment by kaikreuzer** [comment]
> We should try to convince @kaikreuzer to release SAT 0.10.0 after also merging my 2 open PRs of course. 😉

Will try to do so this weekend!

**Comment by J-N-K** [comment]
I excluded test-resources in the XML formatter, they are most likely real data from devices and some tests are sensitive to formatting.

**Comment by J-N-K** [comment]
@wborn What shall we do about the `WebsocketConnection`? It seems like git is treating the file as binary and does not replace the CRLF with LF during upload. Shall we check what happens if we remove the binary data from that file?

**Comment by wborn** [comment]
I'm fine with that for now. Adding a .gitattributes file might also help so you can specify the content and types of line endings for files.

**Comment by J-N-K** [comment]
Ok. Once SAT is released, I'll rebase, apply it again and then we should merge ASAP to avoid new conflicts

**Comment by kaikreuzer** [comment]
Hm, didn't we forget something, guys?
I just release SAT 0.10.0, but it includes https://github.com/openhab/static-code-analysis/pull/371 and https://github.com/openhab/static-code-analysis/pull/373, where we explicitly stated that it "won't be compatible with OH 2.5.x anymore". I hence do not see how it will help us here...?

**Comment by J-N-K** [comment]
We only upgrade the spotless codestyle definitions for SAT, SAT itself stays at 0.9.0 for addons.

**Comment by wborn** [comment]
Thanks for the new SAT release!
I've created https://github.com/openhab/openhab-core/pull/1446 to start using it and also applied/enabled Spotless.

**Comment by J-N-K** [comment]
Please do not push to this PR. This is only for the spotless-changes. We can change other issues later.

**Comment by J-N-K** [comment]
There is our Travis problem again. Would it help to add the `-U` flag to the build command to force re-resolution?

**Comment by J-N-K** [comment]
Seems there are still line-ending issue. I'll check that

**Comment by J-N-K** [comment]
It seems we have problem with Travis and full builds if spotless is enabled. Let's wait what Jenkins has to say. During "mvn spotless:apply" I also had a OOM locally.

**Comment by J-N-K** [comment]
Looks like Jenkins runs into the same memory issue. How much memory memory does the instance have? I ran I to similar problems with 8 GB on a VM.

**Comment by cpmeister** [comment]
Is that OOM a result of a one-time formating change or would it happen even after the formatting changes are made?

**Comment by J-N-K** [comment]
Locally I have it when I use `mvn spotless:apply` on the whole project. But Jenkins and Travis seem to have it during check, too.

**Comment by J-N-K** [comment]
I would suggest to disable the check in the build for now and just commit like it is now. That at least fixed our line-ending problems and other codestyle issues. We can run another „apply“ later after we found the reason for the build failing.

**Comment by J-N-K** [comment]
I also fixed the illegal character issue which led git think that this is a binary file and commit/checkout wrong line endings.

**Comment by wborn** [comment]
Is it ready to be merged now? Jenkins/Travis seem to be happy.

**Comment by J-N-K** [comment]
I think so. The check is disabled, but if we take care, there should be no bigger changes in the future. I‘ll investigate the OOM

**Comment by J-N-K** [comment]
No, my local OOM where at different projects during apply and sometimes (not always) after the normal build was finished (maven already displayed the summary when it happened).

File: bundles/org.openhab.binding.amazonechocontrol/src/main/java/org/openhab/binding/amazonechocontrol/internal/WebSocketConnection.java [hunk 3]
```diff
@@ -555,7 +553,7 @@ void encode(byte[] data, long b, int offset, int len) {
         }
 
         byte[] encodePing() {
-            // MSG 0x00000065 0x0e414e47 f 0x00000001 0xbc2fbb5f 0x00000062 PIN    � ��� ��    R e g u l a rFABE
+            // MSG 0x00000065 0x0e414e47 f 0x00000001 0xbc2fbb5f 0x00000062
```
Comment (wborn) [inline bundles/org.openhab.binding.amazonechocontrol/src/main/java/org/openhab/binding/amazonechocontrol/internal/WebSocketConnection.java:556]: Please remove the trailing space or the check will fail.

```suggestion
            // MSG 0x00000065 0x0e414e47 f 0x00000001 0xbc2fbb5f 0x00000062
```

```
[ERROR] Failed to execute goal com.diffplug.spotless:spotless-maven-plugin:1.30.0:check (default-cli) on project 
org.openhab.binding.amazonechocontrol: The following files had format violations:
[ERROR]     src/main/java/org/openhab/binding/amazonechocontrol/internal/WebSocketConnection.java
[ERROR]         @@ -553,7 +553,7 @@
[ERROR]          ········}
[ERROR]          
[ERROR]          ········byte[]·encodePing()·{
[ERROR]         -············//·MSG·0x00000065·0x0e414e47·f·0x00000001·0xbc2fbb5f·0x00000062·
[ERROR]         +············//·MSG·0x00000065·0x0e414e47·f·0x00000001·0xbc2fbb5f·0x00000062
[ERROR]          ············this.messageId++;
[ERROR]          ············String·msg·=·"MSG·0x00000065·";·//·Message-type·and·Channel·=·CHANNEL_FOR_HEARTBEAT;
[ERROR]          ············msg·+=·this.encodeNumber(this.messageId)·+·"·f·0x00000001·";
[ERROR] Run 'mvn spotless:apply' to fix these violations.
```

**Comment by wborn** [comment]
> No, my local OOM where at different projects during apply and sometimes (not always) after the normal build was finished (maven already displayed the summary when it happened)

Do you build with SAT enabled? It will still be generating the SAT summary reports after the build is finished using this Maven extension:

https://github.com/openhab/openhab-addons/blob/c5fdd32420f98af893bba13c1be39e0f4d5e92eb/pom.xml#L664-L670

You could try what happens if you remove it, disable SAT or increase your memory in the `MAVEN_OPTS`. I have plenty of RAM so I always use: `-Xmx4g`

**Comment by J-N-K** [comment]
@wborn, done.

**Comment by lolodomo** [comment]
This PR is really not cool, it will introduce conflicts probably in most of pending PR !

**Comment by wborn** [comment]
After some investigation I also ran into these out of memory issues when enabling the Spotless checks @J-N-K .

Looks like both the karaf-maven-plugin and spotless-maven-plugin are leaking threads/memory.


When Spotless checks are disabled and the karaf-maven-plugin execution is removed from the [bundle POM](https://github.com/openhab/openhab-addons/blob/7628135323df6f07e3c2993d7b816a16e334d849/bundles/pom.xml#L399) all issues seem to be resolved:

normal

The higher thread counts at the end are due to executing the itests.

---

After enabling the karaf-maven-plugin again, many threads remain started for no good reason:

karaf

---

When the Spotless checks are also enabled, thousands of threads remain started causing some serious memory leakage:

both



The leakage probably scales linear with the number of Maven projects we have in the reactor.

---

Karaf 4.2.8 also seems to have these issues. I didn't find any open issue for this in their issue trackers.

**Comment by J-N-K** [comment]
We should probably file issues for that. We have around 260 projects, so the spotless Plugin created around 18 daemon threads per project. How do you create these graphs? I would try to narrow down what is causing that.

**Comment by wborn** [comment]
I always use [VisualVM](https://visualvm.github.io/) to for analyzing issues with Java. You can connect it to any Java process running on your PC without having to configure anything special. It can also make thread dumps and heap dumps.

This is the full view of the build without those two plugins:

good

The memory graph suggests that probably ~1GB should be enough when building when the leaks are gone.

---


This is the full view when both problem causing plugins are enabled.

bad

It even went out of memory here (4GB limit). You can see that the blue GC activity spiked at the end. Also notice how many additional classes are loaded and that the graph keeps increasing forever. This may indicate there is a classloader leak.

**Comment by J-N-K** [comment]
@lolodomo Unfortunately this is true at any point in time. We'll always have open PR. Most merge conflicts should be solved if you use "mvn spotless:apply" on your branch.

**Comment by lolodomo** [comment]
> @lolodomo Unfortunately this is true at any point in time. We'll always have open PR. Most merge conflicts should be solved if you use "mvn spotless:apply" on your branch.

Good to know for the next time. I will try that on my SamsungTV PR.
For the PR relative to the hue binding, I created a new branch, restored all my changed files and run a git push force.

**Comment by mhilbush** [comment]
So, what about my question [here](https://github.com/openhab/openhab-addons/pull/7449#issuecomment-618037000)?

After rebasing my squeezebox PR to pull in the formatting changes from this PR, using the formatter in Eclipse (Shift-Ctrl-f) changes the formatting from what spotless did.

Is there something that needs to be fixed with my Eclipse config? Does this mean we should no longer use the Eclipse formatter for XML files? Should I always run `mvn spotless:apply` whenever I touch an XML file? What's the recommendation?

**Comment by wborn** [comment]
It should be using the Eclipse WTP formatter for XML files (see [Spotless docs](https://github.com/diffplug/spotless/tree/master/plugin-maven#applying-eclipse-wtp-to-css--html--etc) and our [POM](https://github.com/openhab/openhab-addons/blob/246633607ef19555f8988973abdf7c9797ef0fda/pom.xml#L542-L565)).

The specific XML formatting configuration we use is defined in this [openhab_wst_xml_files.prefs](https://github.com/openhab/static-code-analysis/blob/master/codestyle/src/main/resources/openhab_wst_xml_files.prefs) file.

You can find a similar configuration file in your Eclipse workspace configuration that should match it `.metadata/.plugins/org.eclipse.core.runtime/.settings/org.eclipse.wst.xml.core.prefs`

The preferred way of editting it is using the UI. This is what I see:

xml

I haven't checked if the default openHAB Eclipse IDE settings are the same. If they are different, we should update them. It's also possible there are formatting differences because Spotless is using a different Eclipse formatter version compared to the one used in the Eclipse IDE.

Do you have an example of a file that is formatted in different ways?

**Comment by mhilbush** [comment]
Here's an example of a diff after running the Eclipse formatter.

```
diff --git a/bundles/org.openhab.binding.squeezebox/src/main/resources/ESH-INF/thing/thing-types.xml b/bundles/org.openhab.binding.squeezebox/src/main/resources/ESH-INF/thing/thing-types.xml
index ce1f9bd36..fe9b2ac5a 100644
--- a/bundles/org.openhab.binding.squeezebox/src/main/resources/ESH-INF/thing/thing-types.xml
+++ b/bundles/org.openhab.binding.squeezebox/src/main/resources/ESH-INF/thing/thing-types.xml
@@ -1,6 +1,5 @@
 <?xml version="1.0" encoding="UTF-8"?>
-<thing:thing-descriptions bindingId="squeezebox"
-	xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+<thing:thing-descriptions bindingId="squeezebox" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
 	xmlns:thing="https://openhab.org/schemas/thing-description/v1.0.0"
 	xsi:schemaLocation="https://openhab.org/schemas/thing-description/v1.0.0 https://openhab.org/schemas/thing-description-1.0.0.xsd">
 
@@ -9,7 +8,7 @@
 		<description>This is a SqueezeBox Server instance.</description>
 
 		<channels>
-			<channel id="favoritesList" typeId="favoritesList"/>
+			<channel id="favoritesList" typeId="favoritesList" />
 		</channels>
 
 		<config-description>
@@ -47,42 +46,42 @@
 
 	<thing-type id="squeezeboxplayer">
 		<supported-bridge-type-refs>
-			<bridge-type-ref id="squeezeboxserver"/>
+			<bridge-type-ref id="squeezeboxserver" />
 		</supported-bridge-type-refs>
 
 		<label>SqueezeBox Player</label>
 		<description>This is a SqueezeBox Player instance</description>
 
 		<channels>
-			<channel id="power" typeId="power"/>
-			<channel id="mute" typeId="mute"/>
-			<channel id="volume" typeId="volume"/>
-			<channel id="stop" typeId="stop"/>
-			<channel id="playPause" typeId="playPause"/>
-			<channel id="next" typeId="next"/>
-			<channel id="prev" typeId="prev"/>
-			<channel id="control" typeId="control"/>
-			<channel id="stream" typeId="stream"/>
-			<channel id="source" typeId="source"/>
-			<channel id="sync" typeId="sync"/>
-			<channel id="unsync" typeId="unsync"/>
-			<channel id="playListIndex" typeId="playListIndex"/>
-			<channel id="currentPlayingTime" typeId="currentPlayingTime"/>
-			<channel id="duration" typeId="duration"/>
-			<channel id="currentPlaylistShuffle" typeId="currentPlaylistShuffle"/>
-			<channel id="currentPlaylistRepeat" typeId="currentPlaylistRepeat"/>
-			<channel id="title" typeId="title"/>
-			<channel id="remotetitle" typeId="remotetitle"/>
-			<channel id="album" typeId="album"/>
-			<channel id="artist" typeId="artist"/>
-			<channel id="year" typeId="year"/>
-			<channel id="genre" typeId="genre"/>
-			<channel id="coverartdata" typeId="coverartdata"/>
-			<channel id="ircode" typeId="ircode"/>
-			<channel id="numberPlaylistTracks" typeId="numberPlaylistTracks"/>
-			<channel id="playFavorite" typeId="playFavorite"/>
-			<channel id="like" typeId="like"/>
-			<channel id="unlike" typeId="unlike"/>
+			<channel id="power" typeId="power" />
+			<channel id="mute" typeId="mute" />
+			<channel id="volume" typeId="volume" />
+			<channel id="stop" typeId="stop" />
+			<channel id="playPause" typeId="playPause" />
+			<channel id="next" typeId="next" />
+			<channel id="prev" typeId="prev" />
+			<channel id="control" typeId="control" />
+			<channel id="stream" typeId="stream" />
+			<channel id="source" typeId="source" />
+			<channel id="sync" typeId="sync" />
+			<channel id="unsync" typeId="unsync" />
+			<channel id="playListIndex" typeId="playListIndex" />
+			<channel id="currentPlayingTime" typeId="currentPlayingTime" />
+			<channel id="duration" typeId="duration" />
+			<channel id="currentPlaylistShuffle" typeId="currentPlaylistShuffle" />
+			<channel id="currentPlaylistRepeat" typeId="currentPlaylistRepeat" />
+			<channel id="title" typeId="title" />
+			<channel id="remotetitle" typeId="remotetitle" />
+			<channel id="album" typeId="album" />
+			<channel id="artist" typeId="artist" />
+			<channel id="year" typeId="year" />
+			<channel id="genre" typeId="genre" />
+			<channel id="coverartdata" typeId="coverartdata" />
+			<channel id="ircode" typeId="ircode" />
+			<channel id="numberPlaylistTracks" typeId="numberPlaylistTracks" />
+			<channel id="playFavorite" typeId="playFavorite" />
+			<channel id="like" typeId="like" />
+			<channel id="unlike" typeId="unlike" />
 		</channels>
 
 		<properties>
```
And here's what I see in my Eclipse preferences. "Join line" is the only difference I see.

eclipse-prefs

**Comment by mhilbush** [comment]
If I untick "Insert whitespace ..." and "join lines", then the only diff is this. Do you know how to fix this?

```
diff --git a/bundles/org.openhab.binding.squeezebox/src/main/resources/ESH-INF/thing/thing-types.xml b/bundles/org.openhab.binding.squeezebox/src/main/resources/ESH-INF/thing/thing-types.xml
index ce1f9bd36..2235cd658 100644
--- a/bundles/org.openhab.binding.squeezebox/src/main/resources/ESH-INF/thing/thing-types.xml
+++ b/bundles/org.openhab.binding.squeezebox/src/main/resources/ESH-INF/thing/thing-types.xml
@@ -1,6 +1,5 @@
 <?xml version="1.0" encoding="UTF-8"?>
-<thing:thing-descriptions bindingId="squeezebox"
-	xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+<thing:thing-descriptions bindingId="squeezebox" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
 	xmlns:thing="https://openhab.org/schemas/thing-description/v1.0.0"
 	xsi:schemaLocation="https://openhab.org/schemas/thing-description/v1.0.0 https://openhab.org/schemas/thing-description-1.0.0.xsd">
 ```

**Comment by robnielsen** [comment]
@mhilbush, you also have **Insert whitespace before closing empty end-tags** checked which @wborn  doesn't

**Comment by mhilbush** [comment]
Thanks. Yes, I just caught that. 

See my comment above. After fixing that, I'm down to one line that's different.

**Comment by nedtwigg** [comment]
You can import the eclipse formatter settings along these lines: https://github.com/diffplug/spotless/blob/master/ECLIPSE_SCREENSHOTS.md

**Comment by wborn** [comment]
> If I untick "Insert whitespace ..." and "join lines", then the only diff is this. Do you know how to fix this?

For me it formats that XML file in the Eclipse IDE exactly like how it's formatted by Spotless.

My Eclipse version is:

Version: 2020-03 (4.15.0)
Build id: 20200313-1211

What Eclipse version do you use?

**Comment by mhilbush** [comment]
It's really old, but it works (except for this one formatting issue). Having used it to write 11 bindings, I was comfortable with it. And, having read numerous remarks about the challenges of getting new installs working, I've been putting off upgrading for a long time. I guess it's time to upgrade. 😦

**Comment by wborn** [comment]
> We should probably file issues for that.

There's now also an issue for the karaf-maven-plugin verify goal leaking threads:

https://issues.apache.org/jira/browse/KARAF-6697

**Comment by nedtwigg** [comment]
@mhilbush fwiw, a big part of spotless is that you don't have to worry about how your IDE formats.  Just run `mvn spotless:apply` and it will fix the formatting, so it doesn't matter how your IDE is setup.

**Comment by wborn** [comment]
Are there ways for improving the Spotless check performance @nedtwigg?

Looks like our Jenkins build times increase more than 2.5x by enabling the Spotless check, see:

https://ci.openhab.org/job/PR-openHAB-Core/buildTimeTrend

The check was enabled with https://github.com/openhab/openhab-core/pull/1446 (job `#2195`). As a test I added a commit to https://github.com/openhab/openhab-core/pull/1443 in which the Spotless check is disabled (job `#2213`).

So if we would also enable the check for this repo, build times might also increase 2.5x, i.e. go from ~2 hours to 5+ hours (see [PR build time trend](https://ci.openhab.org/view/Pull%20Request%20Builds/job/PR-openHAB2-Addons/)). 

I hope part of this may also be fixed by the solution for https://github.com/diffplug/spotless/issues/559.

**Comment by wborn** [comment]
I think the decreased performance is mainly due to amount of extra memory being used. We have another build job for that project that's using `-Xmx4096m` instead of `-Xmx768m` which doesn't seem to be have these issues (see [build time trend](https://ci.openhab.org/job/openHAB-Core/buildTimeTrend)). I'll change the `-Xmx` setting of the PR job which hopefully fixes it for now.

**Comment by nedtwigg** [comment]
Indeed, there are multiple big spotless-maven performance problems noted in #559.  There may be other ways to improve, but those are the lowest-hanging fruit

**Comment by andrewfg** [comment]
Has anyone noticed that the Spotless is trashing special characters? For example it trashes the degree symbol characters in my Units-Measure-Code as follows. I think this a breaking change! Or??

**_Before Spotless_**
```
if ("°C".equals(descr)) {
    return SIUnits.CELSIUS;
} else if ("°F".equals(descr)) {
    return ImperialUnits.FAHRENHEIT;
} else if ("K".equals(descr)) {
    return Units.KELVIN;
}
```
**_After Spotless_**
```
if ("ï¿½C".equals(descr)) {
    return SIUnits.CELSIUS;
} else if ("ï¿½F".equals(descr)) {
    return ImperialUnits.FAHRENHEIT;
} else if ("K".equals(descr)) {
    return Units.KELVIN;
}

```

**Comment by nedtwigg** [comment]
Spotless doesn’t do any auto-detection on encoding, it defaults to UTF-8.  You can set to something else (maybe cp1252?) with [this](https://github.com/diffplug/spotless/blob/master/plugin-maven/README.md#line-endings-and-encodings-invisible-stuff).

**Comment by andrewfg** [comment]
@nedtwigg many thanks for the observation. I am sure that there are many ways to change the out of the box settings of MVN spotless, and solve future problems. But that was not my point..

My point was that with the current spotless settings set by the OH team, and the current out of the box settings of Eclipse, the formatter changed the functionality of the above piece of code; (not only how it looks, but also how it works).

_This means that the recent automated application of spotless over the whole existing code-base of OpenHAB **might have broken the existing functionality in some places.**_ - and you guys might not discover it until breaks user's live system.

**Comment by nedtwigg** [comment]
> the formatter changed the functionality of the above piece of code

That is an excellent point!  Of all the times people have used Spotless and had encoding issues, you are actually the first one to correctly report it as a **bug**, which it is.  We've just opened [this PR](https://github.com/diffplug/spotless/pull/575) which will error loudly when/if the encoding is set incorrectly in the future.  It might not have helped with this particular situation, as it seems to have encoded to valid codepoints, just different ones.

Also, these format-the-world commits are currently an unavoidable way to get started with enforced formatting, but they are dangerous.  To fix this, we have the [ratchet feature](https://github.com/diffplug/spotless/issues/511) on the way (hasn't shipped yet) which lets you enforce formatting only on files which have changed since `origin/master` or `v3.1.5`, so that you can adopt auto-formatting change-by-change rather than all at once.

**Comment by nedtwigg** [comment]
To fix your immediate problem, you could:

- checkout the commit before spotless apply
- set encoding correctly
- run spotless apply
- diff the correct-encoding apply with the incorrect-encoding apply

If you then apply the patch from that diff to your current working tree, any regressions will be fixed.  Perhaps a good argument for snapshot-testing that this regression was not caught by a test suite.

**Comment by andrewfg** [comment]
Hi @nedtwigg I have been experimenting, and found out that the best way of eliminating spotless issues seems to be by using inline tags in the source files..
```
//@formatter:off 
..
//@formatter:on

```
I used this to prevent modification of special characters as mentioned already; and in my Junit test class where I have some huge, but nicely formatted, JSON strings for testing the parser. Without applying spotless, the JSON is nicely indented according to elements, sub elements etc. Whereas spotless basically munges the whole string into maximum line length blocks, which looks awful, and is impossible for a human to read..

**Comment by nedtwigg** [comment]
Yes, those tags are useful!  You are using the eclipse formatter which is *highly* configurable.  I agree that it’s a bummer to reorganize existing string literals, I don’t which flag enables that exactly, but I do know that this configuration doesn’t do that:

https://github.com/diffplug/spotless/blob/master/gradle/spotless.eclipseformat.xml

**Comment by andrewfg** [comment]
Crikey! But, usefully the formatter on/off tags have the same impact on Eclipse as they do on Spotless..

**Comment by andrewfg** [comment]
Just to give a final word on this..

For the binding that I am working on, the hardware manufacturer delivers its temperature strings with a degree symbol prefix "°C" which is coded in plain ASCII as decimal code=176 / hex=B0 / binary=1011000 / octal=260.

For whatever historical reason, Java strings allow ASCII characters to be included using **octal** escaping; so the temperature unit string"°C" can be coded as "\260C". 

I have done a few tests before and after spotless, and I can confirm that the code "\260C" is able to "survive" being run through spotless. So in summary, the first assertion below works before spotless, whereas the second assertion below fails because spotless has changed "°" to "ï¿½" 
```
// passes before running spotless
assertEqual("\260C",  "°C"); 

// fails after running spotless, because spotless changed the symbol
assertEqual("\260C",  "ï¿½C"); 
```

**Comment by nedtwigg** [comment]
Minor correction...

> Java strings allow ~~ASCII characters~~ **Unicode codepoints (which may require up to 4 'characters' in UTF-8, or 2 characters in UTF-16)** to be included using octal escaping

In [Cp-1252](https://www.cp1252.com/), '°' encodes to `0xB0`, and `B0` is also the unicode codepoint.  But in UTF-8 (which is Spotless default if you do not set encoding), the `B0` codepoint encodes to `0xC2 0xB0`.  Because Spotless is set to UTF-8, `0xB0` on its own is invalid, and decodes to the codepoint `U+FFFD � REPLACEMENT CHARACTER`, which encodes in UTF-8 to `0xEF 0xBF 0xBD`.  If you read those bytes as Cp-1252:

```
EF BF BD
ï  ¿  ½
```

So it's not randome, it's exactly what you would expect if you roundtripped `Cp-1252`  through `UTF-8`.

**Comment by J-N-K** [comment]
IIRC UTF-8 would be the correct encoding for our files. The exception to this rule are the i18n-files. So we should check which files had the wrong encoding before applying spotless.

**Comment by andrewfg** [comment]
Hi @nedtwigg , I was not suggesting that it is random. My point is that the binding is reading a byte stream over the wire (not a word stream). And in that byte stream are embedded degree symbol characters that are encoded as one single byte xB0. And the binding Java code has to be able to compare these incoming stream strings with a Java string, in order to set the UoM on the data points. My point is that in Eclipse I can also write the Java string with a single byte B0 character, and everything works fine. But when I run spotless over the code, it replaces the single byte character with three bytes. So that when I recompile it, the binding does not work any more. I am just suggesting that you need to find a solution where spotless does not break the functionality of a binding that uses single byte special characters.

**Comment by nedtwigg** [comment]
@andrewfg Agreed, and I appreciate that you have correctly identified this as a bug in Spotless.  Many have had encoding issues, you are the first to correctly identify it as such. 
 https://github.com/diffplug/spotless/pull/575 will fix this for most scenarios (including this one).

**Comment by wborn** [comment]
Doesn't seem to me that there were unwanted file encoding changes (depending on how well `file -i` detects these):

```shell
$ git checkout c5fdd32420f98af893bba13c1be39e0f4d5e92eb
$ find . -type f | xargs -i@ file -i @ | awk -F ':' '{print $2, $1}' | awk '{print $2, $3}' | sort > ../encodings-before
$ git checkout 7628135323df6f07e3c2993d7b816a16e334d849
$ find . -type f | xargs -i@ file -i @ | awk -F ':' '{print $2, $1}' | awk '{print $2, $3}' | sort > ../encodings-after
$ diff ../encodings-before ../encodings-after 
2d1
< charset=binary ./bundles/org.openhab.binding.amazonechocontrol/src/main/java/org/openhab/binding/amazonechocontrol/internal/WebSocketConnection.java
12049a12049
> charset=us-ascii ./bundles/org.openhab.binding.amazonechocontrol/src/main/java/org/openhab/binding/amazonechocontrol/internal/WebSocketConnection.java
20574a20575
> charset=us-ascii ./.gitattributes
```

The WebSocketConnection we discussed in https://github.com/openhab/openhab-addons/pull/7449#issuecomment-619197113.

You can also easily scroll through the big changes of that PR on the CLI with:

`git show 7628135323df6f07e3c2993d7b816a16e334d849`

When I reapply Spotless on c5fdd32420f98af893bba13c1be39e0f4d5e92eb (on Linux), then rebase the result on 7628135323df6f07e3c2993d7b816a16e334d849 there aren't any major differences as well.

**Pull request merged**